source/Mlos.Python/mlos/Optimizers/OptimizerBase.py

```diff
@@ -65,24 +73,96 @@ def optimum(self, stay_focused=False) -> Tuple[Point, Point]: # pylint: disable=
+
+        if optimum_definition == OptimumDefinition.BEST_OBSERVATION:
+            return self._best_observation_optimum(features_df=features_df, objectives_df=objectives_df)
+        return self._prediction_based_optimum(features_df=features_df, optimum_definition=optimum_definition, alpha=alpha)
+
+
+    @trace()
+    def _best_observation_optimum(self, features_df: pd.DataFrame, objectives_df: pd.DataFrame) -> Tuple[Point, Point]:
+        objective = self.optimization_problem.objectives[0]
+        if objective.minimize:
+            index_of_best = objectives_df[objective.name].idxmin()
+        else:
+            index_of_best = objectives_df[objective.name].idxmax()
+        optimum_value = Point.from_dataframe(objectives_df.loc[[index_of_best]])
+        config_at_optimum = Point.from_dataframe(features_df.loc[[index_of_best]])
+        return config_at_optimum, optimum_value
+
+    @trace()
+    def _prediction_based_optimum(self, features_df: pd.DataFrame, optimum_definition: OptimumDefinition, alpha: float)-> Tuple[Point, Point]:
+        objective = self.optimization_problem.objectives[0]
+
+        # Let's see if we have them cached before recomputing
+        #
+        if self._cached_predictions_for_observations is None:
+            self._cached_predictions_for_observations = self.predict(feature_values_pandas_frame=features_df)
+        predictions_df = self._cached_predictions_for_observations.get_dataframe()
+
+        if len(predictions_df.index) == 0:
+            raise ValueError("Insufficient data to compute confidence-bound based optimum.")
+
+        # Predictions index must be a subset of features index.
+        #
+        assert features_df.index.intersection(predictions_df.index).equals(predictions_df.index)
+
+        predicted_value_column_name = Prediction.LegalColumnNames.PREDICTED_VALUE.value
+        dof_column_name = Prediction.LegalColumnNames.DEGREES_OF_FREEDOM.value
+        variance_column_name = Prediction.LegalColumnNames.PREDICTED_VALUE_VARIANCE.value
+
+        if optimum_definition == OptimumDefinition.PREDICTED_VALUE_FOR_OBSERVED_CONFIG:
+            if objective.minimize:
+                index_of_best = predictions_df[predicted_value_column_name].idxmin()
+            else:
+                index_of_best = predictions_df[predicted_value_column_name].idxmax()
+
+            optimum_value = Point.from_dataframe(predictions_df.loc[[index_of_best], [predicted_value_column_name]])
 
-        if self.optimization_problem.objectives[0].minimize:
-            index_of_best_target = objectives_df.idxmin()[0]
         else:
-            index_of_best_target = objectives_df.idxmax()[0]
-        best_objective = Point.from_dataframe(objectives_df.loc[[index_of_best_target]])
-        best_config_point = Point.from_dataframe(features_df.loc[[index_of_best_target]])
+            # We will be manipulating this data so let's make a copy for now, and later optimize this away.

```

Comment:
Maybe mark that with a TODO.  Could be a good "first issue" for someone to tackle later on.

Reply:
Done